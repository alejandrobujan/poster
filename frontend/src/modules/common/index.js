export {default as Pager} from "./components/Pager";
export {default as Errors} from "./components/Errors";
export {default as BackLink} from "./components/BackLink";
export {default as UserCard} from "./components/UserCard";
export {default as PostLink} from "./components/PostLink";
export {default as ImagesCard} from "./components/ImagesCard";