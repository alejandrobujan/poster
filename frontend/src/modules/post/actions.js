import backend from '../../backend';
import * as actionTypes from './actionTypes';
import * as selectors from './selectors';

const findAllCategoriesCompleted = categories => ({
	type: actionTypes.FIND_ALL_CATEGORIES_COMPLETED,
	categories
});

export const findAllCategories = () => (dispatch, getState) => {

	const categories = selectors.getCategories(getState());

	if (!categories) {

		backend.postService.findAllCategories(
			categories => dispatch(findAllCategoriesCompleted(categories))
		);

	}

}

export const createPost = (title, description, url, price, categoryId, images, type, properties,
	onSuccess, onErrors) => dispatch =>
		backend.postService.createPost(title, description, url, price, categoryId, images, type, properties, onSuccess, onErrors);

const findPostsCompleted = postSearch => ({
	type: actionTypes.FIND_POSTS_COMPLETED,
	postSearch
});


const clearPostSearch = () => ({
	type: actionTypes.CLEAR_POST_SEARCH
});


export const findPosts = criteria => dispatch => {

	dispatch(clearPostSearch());
	backend.postService.findPosts(criteria, result =>
		dispatch(findPostsCompleted({ criteria, result }))
	);

};

export const previousFindPostsResultPage = criteria =>
	findPosts({ page: criteria.page - 1 });

export const nextFindPostsResultPage = criteria =>
	findPosts({ page: criteria.page + 1 });

export const findPostById = id => dispatch => {
	backend.postService.findPostById(id,
		post => dispatch(findPostByIdCompleted(post)));
}

const findPostByIdCompleted = post => ({
	type: actionTypes.FIND_POST_BY_ID_COMPLETED,
	post
});

export const clearPost = () => ({
	type: actionTypes.CLEAR_POST
}); 	

export const deletePost = (id, onSuccess, onErrors) => dispatch => {
	backend.postService.deletePost(id, onSuccess, onErrors);
};


export const updatePostCompleted = post => ({
    type: actionTypes.UPDATE_POST_COMPLETED,
    post
})
    
export const updatePost = (post, onSuccess, onErrors) => dispatch =>
	backend.postService.updatePost(post,
		post => {
			dispatch(updatePostCompleted(post));
			onSuccess();
		}, 
		onErrors);

const markPostAsExpiredCompleted = ({expired}) => ({
	type: actionTypes.MARK_POST_AS_EXPIRED_COMPLETED,
	expired
});

export const markPostAsExpired = (id, expired, onErrors) => dispatch => {
	backend.postService.maskPostAsExpired(id, expired, e => dispatch(markPostAsExpiredCompleted(e)), onErrors);
};	

const ratePostCompleted = ratingCount => ({
	type: actionTypes.RATE_POST_COMPLETED,
	ratingCount
});

export const ratePostPositive = (id, onErrors) => dispatch => {
	backend.ratingService.ratePostPositive(id, ratingCount => dispatch(ratePostCompleted(ratingCount)), onErrors);
};


export const ratePostNegative = (id, onErrors) => dispatch => {
	backend.ratingService.ratePostNegative(id, ratingCount => dispatch(ratePostCompleted(ratingCount)), onErrors);
};
