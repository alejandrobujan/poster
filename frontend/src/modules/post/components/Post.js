import ImageGallery from "react-image-gallery";
// import stylesheet if you're not already using CSS @import
import "react-image-gallery/styles/css/image-gallery.css";

const Post = ({ title, description, url, price, categoryDto, images }) => {
	return (
		<div className="col mb-5">
			<div className="card h-100">
				<div className="card-body p-4">
					<div className="text-center">
						<div className="fw-bolder pb-2">
							{categoryDto.name ? categoryDto.name : 'Uncategorized'}
						</div>
						<div className="border-top pt-4">
							{images.length === 0 ?
								(<ImageGallery showFullscreenButton={false} showPlayButton={false} showNav={false}
									items={[{ original: '/poster/assets/noimage.png', originalHeight: "300px" }]} />) :
								(<ImageGallery showFullscreenButton={false} showPlayButton={false}
									items={images.map((image) => ({ original: `data:image/*;base64,${image}`, originalHeight: "300px" }))} />)}
						</div>
						<hr className="my-4" />
						<h5 className="fw-bolder">
							{title}
						</h5>
						<p><i>{description}</i></p>
						<hr className="my-4" />
						<ul className="list-unstyled">

							<li>{url}</li>
							<li>{price}</li>
						</ul>
					</div>
				</div>
			</div>
		</div>);
}

export default Post;