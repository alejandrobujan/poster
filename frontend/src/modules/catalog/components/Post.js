import ImageGallery from "react-image-gallery";
// import stylesheet if you're not already using CSS @import
import "react-image-gallery/styles/css/image-gallery.css";
import {PostLink} from "../../common";
import { ExternalLinkButton } from "../../common";
import OfferIcon from "./OfferIcon";
import CouponIcon from "./CouponIcon";

const Post = ({ id, title, description, url, price, categoryDto, images, type }) => {
	const components = {
		'Offer': OfferIcon,
		'Coupon': CouponIcon
	};
	const Icon = components[type];

return (
		<div className="col mb-5">
			<div className="card h-100">
				<div className="card-body p-4">
					<div className="text-center">
						<div className="fw-bolder pb-2 row">
							<div className="border-right col"><Icon/></div> <div className="col">{categoryDto.name ? categoryDto.name : 'Uncategorized'}</div>
						</div>
						<div className="border-top pt-4">
							{images.length === 0 ?
								(<ImageGallery showFullscreenButton={false} showPlayButton={false} showNav={false}
									items={[{ original: '/poster/assets/noimage.png', originalHeight: "200px" }]} />) :
								(<ImageGallery showFullscreenButton={false} showPlayButton={false}
									items={images.map((image) => ({ original: `data:image/*;base64,${image}`, originalHeight: "200px" }))} />)}
						</div>
						<hr className="my-4" />
						<h5 className="fw-bolder">
							<PostLink id={id} title={title}/> &nbsp; {url &&
								<ExternalLinkButton url={`//${url}`} />
							}
						</h5>
						<p><i>{description}</i></p>

						<hr className="my-4" />
						<p className="post-price">{price.toFixed(2)}â‚¬</p>
					</div>
				</div>
			</div>
		</div>);
	
}

export default Post;