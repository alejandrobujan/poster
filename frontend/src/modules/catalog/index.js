import * as actions from './actions';
import * as actionTypes from './actionTypes';
import reducer from './reducer';
import * as selectors from './selectors';

export {default as SearchBar} from "./components/SearchBar";
export {default as CategorySelector} from './components/CategorySelector';
export {default as Post} from './components/Post';
export {default as Posts} from './components/Posts';
export {default as Feed} from './components/Feed';

// eslint-disable-next-line
export default {actions, actionTypes, reducer, selectors};


