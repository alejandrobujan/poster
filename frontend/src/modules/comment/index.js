import * as actions from './actions';
import reducer from './reducer';
import * as selectors from './selectors';

export {default as Comments} from './components/Comments';
export {default as Comment} from './components/Comment';

// eslint-disable-next-line
export default  {actions, reducer, selectors};