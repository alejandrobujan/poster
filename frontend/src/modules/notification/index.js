import * as actions from './actions';
import * as actionTypes from './actionTypes';
import reducer from './reducer';
import * as selectors from './selectors';

export {default as Notifications} from "./components/Notifications";
export {default as Notification} from "./components/Notification";

// eslint-disable-next-line
export default {actions, actionTypes, reducer, selectors};


