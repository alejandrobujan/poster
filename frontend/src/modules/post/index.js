import * as actions from './actions';
import reducer from './reducer';
import * as selectors from './selectors';

export {default as CreatePost} from './components/CreatePost';
export {default as CategorySelector} from './components/CategorySelector';
export {default as Post} from './components/Post';
export {default as Posts} from './components/Posts';
export {default as Feed} from './components/Feed';
export {default as PostDetails} from './components/PostDetails';


// eslint-disable-next-line
export default  {actions, reducer, selectors};