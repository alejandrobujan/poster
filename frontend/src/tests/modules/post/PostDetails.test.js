import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from 'react-redux';

import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import renderer from "react-test-renderer";
import { postMock } from "../../state/Post.mock";

import { PostDetails } from "../../../modules/post";

describe("PostDetails", () => {
	const middlewares = [thunk];
	const mockStore = configureMockStore(middlewares);
	let store;
	store = mockStore(postMock);

	it("renders correctly", () => {
		const tree = renderer
			.create(
				<Provider store={store}>
					<MemoryRouter>
						<PostDetails />
					</MemoryRouter>
				</Provider>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it("basic structure", () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
					<PostDetails />
				</MemoryRouter>
			</Provider>
		);

		const commentForm = screen.getByTestId('CommentForm');
		const comments = screen.getByTestId('Comments');
	});

	it("create comment", async () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
					<PostDetails />
				</MemoryRouter>
			</Provider>
		);

		const inputComment = screen.getByTestId('CommentInput');

		fireEvent.change(inputComment, {
			target: { value: "Hola, buenos dias" }
		});

		const commentButton = screen.getByTestId('CommentButton');

		fireEvent.click(commentButton);

	});

	it("reply comment empty", async () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
					<PostDetails />
				</MemoryRouter>
			</Provider>
		);



		const inputComment = screen.getByTestId('CommentInput');

		fireEvent.change(inputComment, {
			target: { value: "" }
		});

		const commentButton = screen.getByTestId('CommentButton');

		fireEvent.click(commentButton);

		const errorInput = screen.getByTestId('ErrorForm');

	});
	
	it("click mark post as valid", async () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
                    <PostDetails />
				</MemoryRouter>
			</Provider>
		);

        const markAsValidButton = screen.getByTestId('MarkAsValidButton');

        await waitFor(() => fireEvent.click(markAsValidButton));  
        
        const markedAsValidText = screen.getByTestId('MarkedAsValidText');        
            
	});
	
	it("click save post", async () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
                    <PostDetails />
				</MemoryRouter>
			</Provider>
		);

        const savePostButton = screen.getByTestId('SavePostButton');

        await waitFor(() => fireEvent.click(savePostButton));     
            
	});




});