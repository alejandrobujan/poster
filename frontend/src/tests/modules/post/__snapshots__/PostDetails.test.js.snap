// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostDetails renders correctly 1`] = `
<div
  className="container mt-5"
>
  <div
    className="row"
  >
    <div
      className="col-lg-6 text-center"
    >
      <div
        aria-live="polite"
        className="image-gallery"
      >
        <div
          className="image-gallery-content  image-gallery-thumbnails-bottom"
        >
          <div
            className="image-gallery-slide-wrapper  image-gallery-thumbnails-bottom"
          >
            <div
              className="image-gallery-slides"
            >
              <div
                aria-label="Go to Slide 1"
                className="image-gallery-slide  image-gallery-center "
                onClick={null}
                onFocus={null}
                onKeyUp={[Function]}
                onMouseLeave={null}
                onMouseOver={null}
                onTouchEnd={null}
                onTouchMove={null}
                onTouchStart={null}
                role="button"
                style={
                  Object {
                    "MozTransform": "translate3d(0%, 0, 0)",
                    "OTransform": "translate3d(0%, 0, 0)",
                    "WebkitTransform": "translate3d(0%, 0, 0)",
                    "display": "inherit",
                    "msTransform": "translate3d(0%, 0, 0)",
                    "transform": "translate3d(0%, 0, 0)",
                  }
                }
                tabIndex="-1"
              >
                <img
                  className="image-gallery-image"
                  height="300px"
                  onError={[Function]}
                  onLoad={[Function]}
                  src="/poster/assets/noimage.png"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="mt-4"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
        >
          <div
            className="d-flex align-items-center"
          >
            <svg
              className="bi bi-arrow-left"
              fill="currentColor"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15 8a.5.5 0 0 0-.5-.5H2.707l3.147-3.146a.5.5 0 1 0-.708-.708l-4 4a.5.5 0 0 0 0 .708l4 4a.5.5 0 0 0 .708-.708L2.707 8.5H14.5A.5.5 0 0 0 15 8z"
                fillRule="evenodd"
              />
            </svg>
            <span
              className="ml-2"
            >
              Back
            </span>
          </div>
        </button>
          
        <a
          className="btn btn-primary"
          href="//http://cupon"
          rel="noreferrer"
          target="_blank"
        >
          <div
            className="d-flex align-items-center"
          >
            <svg
              className="bi bi-box-arrow-up-right"
              fill="currentColor"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8.636 3.5a.5.5 0 0 0-.5-.5H1.5A1.5 1.5 0 0 0 0 4.5v10A1.5 1.5 0 0 0 1.5 16h10a1.5 1.5 0 0 0 1.5-1.5V7.864a.5.5 0 0 0-1 0V14.5a.5.5 0 0 1-.5.5h-10a.5.5 0 0 1-.5-.5v-10a.5.5 0 0 1 .5-.5h6.636a.5.5 0 0 0 .5-.5z"
                fillRule="evenodd"
              />
              <path
                d="M16 .5a.5.5 0 0 0-.5-.5h-5a.5.5 0 0 0 0 1h3.793L6.146 9.146a.5.5 0 1 0 .708.708L15 1.707V5.5a.5.5 0 0 0 1 0v-5z"
                fillRule="evenodd"
              />
            </svg>
            <span
              className="ml-2"
            >
              Go to site
            </span>
          </div>
        </a>
         
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          <svg
            className="bi bi-hand-thumbs-up-fill"
            fill="currentColor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.956 1.745C7.021.81 7.908.087 8.864.325l.261.066c.463.116.874.456 1.012.965.22.816.533 2.511.062 4.51a9.84 9.84 0 0 1 .443-.051c.713-.065 1.669-.072 2.516.21.518.173.994.681 1.2 1.273.184.532.16 1.162-.234 1.733.058.119.103.242.138.363.077.27.113.567.113.856 0 .289-.036.586-.113.856-.039.135-.09.273-.16.404.169.387.107.819-.003 1.148a3.163 3.163 0 0 1-.488.901c.054.152.076.312.076.465 0 .305-.089.625-.253.912C13.1 15.522 12.437 16 11.5 16H8c-.605 0-1.07-.081-1.466-.218a4.82 4.82 0 0 1-.97-.484l-.048-.03c-.504-.307-.999-.609-2.068-.722C2.682 14.464 2 13.846 2 13V9c0-.85.685-1.432 1.357-1.615.849-.232 1.574-.787 2.132-1.41.56-.627.914-1.28 1.039-1.639.199-.575.356-1.539.428-2.59z"
            />
          </svg>
           
          0
        </button>
         
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          <svg
            className="bi bi-hand-thumbs-down-fill"
            fill="currentColor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.956 14.534c.065.936.952 1.659 1.908 1.42l.261-.065a1.378 1.378 0 0 0 1.012-.965c.22-.816.533-2.512.062-4.51.136.02.285.037.443.051.713.065 1.669.071 2.516-.211.518-.173.994-.68 1.2-1.272a1.896 1.896 0 0 0-.234-1.734c.058-.118.103-.242.138-.362.077-.27.113-.568.113-.856 0-.29-.036-.586-.113-.857a2.094 2.094 0 0 0-.16-.403c.169-.387.107-.82-.003-1.149a3.162 3.162 0 0 0-.488-.9c.054-.153.076-.313.076-.465a1.86 1.86 0 0 0-.253-.912C13.1.757 12.437.28 11.5.28H8c-.605 0-1.07.08-1.466.217a4.823 4.823 0 0 0-.97.485l-.048.029c-.504.308-.999.61-2.068.723C2.682 1.815 2 2.434 2 3.279v4c0 .851.685 1.433 1.357 1.616.849.232 1.574.787 2.132 1.41.56.626.914 1.28 1.039 1.638.199.575.356 1.54.428 2.591z"
            />
          </svg>
           
          0
        </button>
         
         
      </div>
    </div>
    <div
      className="col-lg-6 text-left"
    >
      <div>
        <svg
          className="bi bi-ticket-detailed"
          fill="currentColor"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 5.5a.5.5 0 0 1 .5-.5h7a.5.5 0 0 1 0 1h-7a.5.5 0 0 1-.5-.5Zm0 5a.5.5 0 0 1 .5-.5h7a.5.5 0 0 1 0 1h-7a.5.5 0 0 1-.5-.5ZM5 7a1 1 0 0 0 0 2h6a1 1 0 1 0 0-2H5Z"
          />
          <path
            d="M0 4.5A1.5 1.5 0 0 1 1.5 3h13A1.5 1.5 0 0 1 16 4.5V6a.5.5 0 0 1-.5.5 1.5 1.5 0 0 0 0 3 .5.5 0 0 1 .5.5v1.5a1.5 1.5 0 0 1-1.5 1.5h-13A1.5 1.5 0 0 1 0 11.5V10a.5.5 0 0 1 .5-.5 1.5 1.5 0 1 0 0-3A.5.5 0 0 1 0 6V4.5ZM1.5 4a.5.5 0 0 0-.5.5v1.05a2.5 2.5 0 0 1 0 4.9v1.05a.5.5 0 0 0 .5.5h13a.5.5 0 0 0 .5-.5v-1.05a2.5 2.5 0 0 1 0-4.9V4.5a.5.5 0 0 0-.5-.5h-13Z"
          />
        </svg>
         Coupon
      </div>
      <hr
        className="my-1"
      />
      <h5
        className="fw-bolder"
      >
        Coupon
      </h5>
      <p
        className="card-text"
      >
        <i>
          Buena oferta
        </i>
      </p>
      <p
        className="post-price"
      >
        10.00
        €
      </p>
      <p
        className="card-text"
      >
        <strong>
          Category:
        </strong>
         
        Belleza
      </p>
      <div
        className="copy-button"
      >
        <input
          id="copyvalue"
          readOnly={true}
          type="text"
          value="12345"
        />
        <button
          className="btn btn-primary"
          onClick={[Function]}
        >
          <svg
            className="bi bi-copy"
            fill="currentColor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 2a2 2 0 0 1 2-2h8a2 2 0 0 1 2 2v8a2 2 0 0 1-2 2H6a2 2 0 0 1-2-2V2Zm2-1a1 1 0 0 0-1 1v8a1 1 0 0 0 1 1h8a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H6ZM2 5a1 1 0 0 0-1 1v8a1 1 0 0 0 1 1h8a1 1 0 0 0 1-1v-1h1v1a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2h1v1H2Z"
              fillRule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        className="post-signature"
      >
        <div
          className="creation-date"
        >
          Posted at 
          1/1/1970 12:00:00 AM
        </div>
        <div
          className="user-card"
        >
          <img
            alt="Avatar"
            height="30px"
            src="/poster/assets/profile.png"
            width="30px"
          />
          <div>
            <div>
              <b>
                Pepe
              </b>
            </div>
            <div>
              <i>
                Pepe Pepe
              </i>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
     
    <form
      className="needs-validation container ml-1"
      data-testid="CommentForm"
      noValidate={true}
      onSubmit={[Function]}
    >
      <div>
        <input
          autoFocus={true}
          className="form-control"
          data-testid="CommentInput"
          id="comment"
          maxLength={256}
          minLength={1}
          onChange={[Function]}
          placeholder="Add a comment"
          required={true}
          type="text"
          value=""
        />
        <div
          className="invalid-feedback"
          data-testid="ErrorForm"
        >
          The comment size must be between 1 and 256
        </div>
      </div>
      <div
        className="text-right"
      >
        <button
          className="btn btn-primary my-3"
          data-testid="CommentButton"
          type="submit"
        >
          Comment
        </button>
      </div>
    </form>
    <div
      data-testid="Comments"
    >
      <div
        className="alert alert-danger"
        data-testid="NoComments"
        role="alert"
      >
        No comments found
      </div>
      <nav
        aria-label="page navigation"
      >
        <ul
          className="pagination justify-content-center"
        >
          <li
            className="page-item disabled"
          >
            <button
              className="page-link"
              onClick={[Function]}
            >
              Back
            </button>
          </li>
          <li
            className="page-item disabled"
          >
            <button
              className="page-link"
              onClick={[Function]}
            >
              Next
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</div>
`;
